flatpak: Support parsing a list of packages. #41996

LukeShortCloud opened the pull request:
##### SUMMARY
Package modules since Ansible 2.7 support using a list instead of a string. This adds support to the flatpak module to install and uninstall a list of packages.

Fixes #41996

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
flatpak module

##### ANSIBLE VERSION
2.8

LukeShortCloud committed:
flatpak: Support parsing a list of packages. #41996

Package modules since Ansible 2.7 support
using a list instead of a string. This
adds support to the flatpak module to install
and uninstall a list of packages.

maxamillion commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -94,6 +94,11 @@
+
+- name: Install a list of flatpaks
+  flatpak:
+    name: ["io.atom.Atom", "org.libreoffice.LibreOffice"]
```
For documentation, we generally prefer the list format:
```
- name: Install a list of flatpaks
  flatpak:
     name:
      - "io.atom.Atom"
      - "org.libreoffice.LibreOffice"
```

LukeShortCloud replied:
I have now updated the example to be a YAML list.

maxamillion commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -132,10 +137,19 @@
+    for app in name:
+
+        if app.startswith('http://') or app.startswith('https://'):
+            use_remote = False
```
What if some apps in the list `name` start with `https://` and some don't, with the `use_remote=False` case those that don't start with `https://` to fail installation? If so, there should be some book keeping done here to handle that.

LukeShortCloud replied:
Thank you for bringing this up! I have rebased my patch and added an update to deal with a mix of URLs and non-URLs (just names) in the list.

Commit:
flatpak: Handle a mix of flatpak URLs and names.

The names list should now properly install
all flatpaks even if some use an explicit
URL to a different remote.

maxamillion commented:
rebuild_merge

maxamillion commented:
bot_status

maxamillion commented:
rebuild_merge

gundalow commented:
@ekultails Could you please rebase so we can get this merged, thanks

oolongbrothers commented:
And please, please, pretty please run the integration tests before merging. The last thing we want is more regressions in the flatpak module. There are already enough of them with fixes that don't get merged.

ekultails commented:
I have rebased my patch set. The integration tests are also succeeding.

```
# cd ansible
# . hacking/env-setup
# ansible-test integration -v --docker fedora28 --docker-privileged --allow-unsupported flatpak
```
```
PLAY RECAP *********************************************************************
testhost                   : ok=63   changed=20   unreachable=0    failed=0    skipped=2    rescued=0    ignored=1
```

oolongbrothers commented:
Great, thanks for running the integration tests for Fedora 28 and Python 2!

However, when I run the integration tests for Fedora 29 and Python 3, I get the following error:
```
ansible-test integration -v --docker fedora29 --docker-privileged --allow-unsupported --python 3.7 flatpak
[...]
TASK [flatpak : Add flatpak for testing check mode on present flatpak] *********
fatal: [testhost]: FAILED! => {"changed": false, "module_stderr": "/root/.ansible/tmp/ansible-tmp-1553098715.4927335-204679751246425/AnsiballZ_flatpak.py:32: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses\n
  import imp\n
Traceback (most recent call last):\n
  File \"/root/.ansible/tmp/ansible-tmp-1553098715.4927335-204679751246425/AnsiballZ_flatpak.py\", line 128, in <module>\n
    _ansiballz_main()\n
  File \"/root/.ansible/tmp/ansible-tmp-1553098715.4927335-204679751246425/AnsiballZ_flatpak.py\", line 120, in _ansiballz_main\n
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n
  File \"/root/.ansible/tmp/ansible-tmp-1553098715.4927335-204679751246425/AnsiballZ_flatpak.py\", line 63, in invoke_module\n
    imp.load_module('__main__', mod, module, MOD_DESC)\n
  File \"/usr/lib64/python3.7/imp.py\", line 234, in load_module\n
    return load_source(name, filename, file)\n
  File \"/usr/lib64/python3.7/imp.py\", line 169, in load_source\n
    module = _exec(spec, sys.modules[name])\n
  File \"<frozen importlib._bootstrap>\", line 630, in _exec\n
  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n
  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n
  File \"/tmp/ansible_flatpak_payload_bwgraq_o/__main__.py\", line 326, in <module>\n
  File \"/tmp/ansible_flatpak_payload_bwgraq_o/__main__.py\", line 312, in main\n
  File \"/tmp/ansible_flatpak_payload_bwgraq_o/__main__.py\", line 200, in flatpak_exists\n
  File \"/tmp/ansible_flatpak_payload_bwgraq_o/__main__.py\", line 263, in _flatpak_command\n
TypeError: can only concatenate str (not \"bytes\") to str\n
", "module_stdout": "", "msg": "MODULE FAILURE\n
See stdout/stderr for the exact error", "rc": 1}
```

Looks like your classic Python 2/3 compatibility problem. I assume you will have to guard your concatenation and possibly other places as well with `to_native` from `module_utils._text`, to assure compatibility, like I did here:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/flatpak.py#L228

oolongbrothers commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -220,19 +259,21 @@ def _flatpak_command(module, noop, command):
-    result['stdout'] = stdout_data
-    result['stderr'] = stderr_data
+    result['command'].append(command)
+    result['stdout'] += stdout_data
```
```suggestion
    result['stdout'] += to_native(stdout_data)
```

oolongbrothers commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -220,19 +259,21 @@ def _flatpak_command(module, noop, command):
-    result['command'] = command
-    result['stdout'] = stdout_data
-    result['stderr'] = stderr_data
+    result['command'].append(command)
```
```suggestion
    result['command'].append(to_native(command))
```

oolongbrothers commented on lib/ansible/modules/packaging/os/flatpak.py:
```diff
@@ -220,19 +259,21 @@ def _flatpak_command(module, noop, command):
-    result['stderr'] = stderr_data
+    result['command'].append(command)
+    result['stdout'] += stdout_data
+    result['stderr'] += stderr_data
```
```suggestion
    result['stderr'] += to_native(stderr_data)
```

oolongbrothers reviewed (changes requested):
These should do it regarding Python 2/3 compatibility.

oolongbrothers commented:
With the changes above applied, the tests run further, but on Fedora 29 then fail with the following error:
```
TASK [flatpak : Test removal - user] *******************************************
fatal: [testhost]: FAILED! => {"changed": false, "command": ["/usr/bin/flatpak list --user --app", "/usr/bin/flatpak list --user --app", "/usr/bin/flatpak uninstall --user -y GNOME"], "flatpaks_found": ["org.gnome.Characters"], "msg": "Failed to execute flatpak command", "rc": 1, "stderr": "error: No ref chosen to resolve matches for ‘GNOME’\n
", "stderr_lines": ["error: No ref chosen to resolve matches for ‘GNOME’"], "stdout": "GNOME Characters - Character map application\torg.gnome.Characters\t\tstable\tx86_64\tflathub\n
GNOME Characters - Character map application\torg.gnome.Characters\t\tstable\tx86_64\tflathub\n
Similar installed refs found for ‘GNOME’:\n
\n
   1) runtime/org.gnome.Platform/x86_64/3.28 (user)\n
   2) app/org.gnome.Characters/x86_64/stable (user)\n
   3) All of the above\n
\n
Which do you want to use (0 to abort)? [0-3]: 0\n
", "stdout_lines": ["GNOME Characters - Character map application\torg.gnome.Characters\t\tstable\tx86_64\tflathub", "GNOME Characters - Character map application\torg.gnome.Characters\t\tstable\tx86_64\tflathub", "Similar installed refs found for ‘GNOME’:", "", "   1) runtime/org.gnome.Platform/x86_64/3.28 (user)", "   2) app/org.gnome.Characters/x86_64/stable (user)", "   3) All of the above", "", "Which do you want to use (0 to abort)? [0-3]: 0"]}
```

That is a bug with flatpak > 1.2.0 compatibility that I fixed in https://github.com/ansible/ansible/pull/51482, which in itself is a substantial fix.
That bug currently blocks us from being able to run all integration tests and hence being able to assess if this code is free of regressions for flatpak >= 1.2.0.

What I think is that we should first merge the bugfix PR #51482 and then rebase the list feature on onto that clean state.
What do you guys think?

ekultails commented:
@oolongbrothers Thanks for the suggestions! I agree, that seems reasonable to wait for your bug fix patch to land first. Then I can rebase this again and test against Python 2 and 3 next time around.

oolongbrothers commented:
@ekultails if you want to accelerate things, you could put your flatpak module experience to use by having a look at said #51482 
That additional review could be all that's needed to get it merged.

LukeShortCloud committed:
flatpak: Add Luke Short (@ekultails) to the list of contributors.

ekultails commented:
I have rebased my patch set ontop of the latest devel. Integration tests are passing on my end, too! Hopefully it is not too late to get this into Ansible 2.8.0. I noticed the first alpha release was tagged yesterday.

oolongbrothers commented:
Awesome, @ekultails !

I can confirm that the integration tests run over all these permutations report no regressions (including both pre and post-1.2 versions of flatpak):
```
ansible-test integration -v --docker fedora28 --docker-privileged --allow-unsupported --python 2.7 flatpak
ansible-test integration -v --docker fedora29 --docker-privileged --allow-unsupported --python 3.7 flatpak
ansible-test integration -v --docker ubuntu1604py2 --docker-privileged --allow-unsupported --python 2.7 flatpak
ansible-test integration -v --docker ubuntu1604py3 --docker-privileged --allow-unsupported --python 3.5 flatpak
ansible-test integration -v --docker ubuntu1804 --docker-privileged --allow-unsupported --python 3.6 flatpak
```

After checking the above, I dusted off some integration tests that I had written earlier for the list feature. Coverage is not 100%, but I think it's good enough. See this commit: https://github.com/oolongbrothers/ansible/commit/56ec72ad905b0d0dfc86d156605762253eb430a3
It would be great if you could add them to your PR. You could just cherry-pick the commit.

The tests confirm, that the list feature works just fine when supplying flatpak name/remote.
There are some issues when supplying lists of `flatpakref`-URLs though:

### Issue 1: Addition with list of URLs

flatpak seems not to support more than one parameter when installing flatpaks via URL: Running `flatpak install --user -y https://flathub.org/repo/appstream/org.gnome.quadrapassel.flatpakref https://flathub.org/repo/appstream/org.gnome.tetravex.flatpakref` results in a `Too many arguments` error.

Test:
```
- name: Test addition with list of urls - user
  flatpak:
    name:
      - https://flathub.org/repo/appstream/org.gnome.quadrapassel.flatpakref
      - https://flathub.org/repo/appstream/org.gnome.tetravex.flatpakref
    state: present
    method: user
  register: url_addition_list_result
```

Full output:
```
$ ansible-test integration -v --docker fedora28 --docker-privileged --allow-unsupported --python 2.7 flatpak
[...]
TASK [flatpak : Test addition with list of urls - user] ************************
fatal: [testhost]: FAILED! => 
{
    "changed": false,
    "command": [
        "/usr/bin/flatpak list --user --app",
        "/usr/bin/flatpak list --user --app",
        "/usr/bin/flatpak install --user -y https://flathub.org/repo/appstream/org.gnome.quadrapassel.flatpakref https://flathub.org/repo/appstream/org.gnome.tetravex.flatpakref"
    ],
    "flatpaks_found": [],
    "msg": "Failed to execute flatpak command",
    "rc": 1,
    "stderr": "Usage:\n  /usr/bin/flatpak install [OPTION…] LOCATION/REMOTE [REF...] - Install applications or runtimes\n\nHelp Options:\n  -h, --help              Show help options\n\nApplication Options:\n  --user                  Work on user installations\n  --system                Work on system-wide installations (default)\n  --installation=NAME     Work on specific system-wide installation(s)\n  --arch=ARCH             Arch to install for\n  --no-pull               Don't pull, only install from local cache\n  --no-deploy             Don't deploy, only download to local cache\n  --no-related            Don't install related refs\n  --no-deps               Don't verify/install runtime dependencies\n  --no-static-deltas      Don't use static deltas\n  --runtime               Look for runtime with the specified name\n  --app                   Look for app with the specified name\n  --bundle                Assume LOCATION is a .flatpak single-file bundle\n  --from                  Assume LOCATION is a .flatpakref application description\n  --gpg-file=FILE         Check bundle signatures with GPG key from FILE (- for stdin)\n  --subpath=PATH          Only install this subpath\n  -y, --assumeyes         Automatically answer yes for all questions\n  --reinstall             Uninstall first if already installed\n  -v, --verbose           Print debug information during command processing, -vv for more detail\n  --ostree-verbose        Print OSTree debug information during command processing\n\nerror: Too many arguments\n",
    "stderr_lines": [
        "Usage:",
        "  /usr/bin/flatpak install [OPTION…] LOCATION/REMOTE [REF...] - Install applications or runtimes",
        "",
        "Help Options:",
        "  -h, --help              Show help options",
        "",
        "Application Options:",
        "  --user                  Work on user installations",
        "  --system                Work on system-wide installations (default)",
        "  --installation=NAME     Work on specific system-wide installation(s)",
        "  --arch=ARCH             Arch to install for",
        "  --no-pull               Don't pull, only install from local cache",
        "  --no-deploy             Don't deploy, only download to local cache",
        "  --no-related            Don't install related refs",
        "  --no-deps               Don't verify/install runtime dependencies",
        "  --no-static-deltas      Don't use static deltas",
        "  --runtime               Look for runtime with the specified name",
        "  --app                   Look for app with the specified name",
        "  --bundle                Assume LOCATION is a .flatpak single-file bundle",
        "  --from                  Assume LOCATION is a .flatpakref application description",
        "  --gpg-file=FILE         Check bundle signatures with GPG key from FILE (- for stdin)",
        "  --subpath=PATH          Only install this subpath",
        "  -y, --assumeyes         Automatically answer yes for all questions",
        "  --reinstall             Uninstall first if already installed",
        "  -v, --verbose           Print debug information during command processing, -vv for more detail",
        "  --ostree-verbose        Print OSTree debug information during command processing",
        "",
        "error: Too many arguments"
    ],
    "stdout": "org.gnome.Characters/x86_64/stable\tcurrent\norg.gnome.Characters/x86_64/stable\tcurrent\n",
    "stdout_lines": [
        "org.gnome.Characters/x86_64/stable\tcurrent",
        "org.gnome.Characters/x86_64/stable\tcurrent"
    ]
}
```

### Issue 2: Removal with list of URLs

Removal with a list of URLs does not guess the flatpak name properly, so no flatpaks are removed.

Test:
```
- name: Test removal with list of urls - user
  flatpak:
    name:
      - https://flathub.org/repo/appstream/org.gnome.quadrapassel.flatpakref
      - https://flathub.org/repo/appstream/org.gnome.tetravex.flatpakref
    state: absent
    method: user
  register: url_removal_list_result
```

Full output:
```
$ ansible-test integration -v --docker fedora28 --docker-privileged --allow-unsupported --python 2.7 flatpak
[...]
TASK [flatpak : Test removal with list of urls - user] *************************
ok: [testhost] => {"changed": false, "command": ["/usr/bin/flatpak list --user --app", "/usr/bin/flatpak list --user --app"], "flatpaks_found": [], "rc": 0, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [flatpak : Verify removal with list of urls test result - user] ***********
fatal: [testhost]: FAILED! => {
    "assertion": "url_removal_list_result.changed == true", 
    "changed": false, 
    "evaluated_to": false, 
    "msg": "state=absent with list of urls as name shall remove flatpaks when present"
}
```

Unfortunately, I didn't have the opportunity to look into fixing these issues, but I am sure that these are just minor issues. 

Thanks a bunch for your effort. We're so close now!

ekultails commented:
Thanks for the feedback! I'll take your suggestions and try to get an updated patch set out soon.
